### PR TITLE
Allow weights in statistical recipes

### PR DESCRIPTION
**Pull request by greimel**
Add the keyword `weights` to `boxplot`, `density`, `hist`, `violin`.

## `density`

image

```julia
let
	N = 100_000
	x = rand(Uniform(-2, 2), N)
	
	w = pdf.(Normal(), x)

	fig = Figure()
	density(fig[1,1], x)
	density(fig[1,2], x, weights = w)

	fig
end
```

## `violin`
image

```julia
let
	N = 100_000
	x = rand(1:3, N)
	y = rand(Uniform(-1, 5), N)
	
	w = pdf.(Normal(), x .- y)

	fig = Figure()

	violin(fig[1,1], x, y)
	violin(fig[1,2], x, y, weights = w)

	fig
end
```

## `boxplot`
image

```julia
let
	N = 100_000
	x = rand(1:3, N)
	y = rand(Uniform(-1, 5), N)
	
	w = pdf.(Normal(), x .- y)

	fig = Figure()

	boxplot(fig[1,1], x, y)
	boxplot(fig[1,2], x, y, weights = w)

	fig
end
```

## `hist`
image
```julia
let
	N = 100_000
	x = rand(Uniform(-5, 5), N)
	
	w = pdf.(Normal(), x)

	fig = Figure()
	hist(fig[1,1], x)
	hist(fig[1,2], x, weights = w)

	fig
end
```

### CODE REVIEW

**Comment by greimel** [comment]
Friendly bump 😃 

The added functionality generalizes what `AlgebraOfGraphics.density()` and `AlgebraOfGraphics.histogram()` can already do. (e.g. `data(df) * mapping(:y, weights = :wgts) * density() |> draw`)

@piever, since you have written most of the code that this PR touches – would you mind reviewing it if you find some time?

File: src/stats/violin.jl [hunk 1]
```diff
@@ -5,6 +5,7 @@ Draw a violin plot.
 - `x`: positions of the categories
 - `y`: variables whose density is computed
 # Keywords
+- `weights`: vector of statistical weights (length of data). By default, each observation has weight 1.
```
Comment (piever) [inline src/stats/violin.jl:8]:
```suggestion
- `weights`: vector of statistical weights (length of data). By default, each observation has weight `1`.
```

File: src/stats/violin.jl [hunk 2]
```diff
@@ -71,6 +73,7 @@ function plot!(plot::Violin)
                 npoints = n,
                 (bound === automatic ? NamedTuple() : (boundary = bound,))...,
                 (bw === automatic ? NamedTuple() : (bandwidth = bw,))...,
+                (w === automatic ? NamedTuple() : (weights = StatsBase.weights(view(w, idxs)),))...
```
Comment (piever) [inline src/stats/violin.jl:76]: We should probably add a utility function to filter out `automatic` keywords, but that should be done separately.

File: src/stats/boxplot.jl [hunk 3]
```diff
@@ -17,6 +17,7 @@ The boxplot has 3 components:
 - `x`: positions of the categories
 - `y`: variables within the boxes
 # Keywords
+- `weights`: vector of statistical weights (length of data). By default, each observation has weight 1.
```
Comment (piever) [inline src/stats/boxplot.jl:20]:
```suggestion
- `weights`: vector of statistical weights (length of data). By default, each observation has weight `1`.
```

**Review by piever — CHANGES_REQUESTED** [review]

**Comment by piever** [comment]
Nice! Looks good to me (just left two tiny formatting comments).

I have a minor doubt as to whether we should call `StatsBase.weights` indiscriminately, or instead not call it when the user passes a `StatsBase.AbstractWeight` object. I have no idea whether the various subtypes of `AbstractWeight` make any difference here. I suspect not, so probably it's fine to leave as is.

**Comment by greimel** [comment]
I've applied your suggestions, thanks for the review!

**Review by piever — APPROVED** [review]

**Comment by SimonDanisch** [comment]
Great!
There is still an error in the docs:
```
┌ Franklin Warning: in <examples/plotting_functions/violin.md>
│ There was an error of type 'ArgumentError' when running a code block.
│ Checking the output files '/home/runner/work/Makie.jl/Makie.jl/docs/__site/assets/examples/plotting_functions/violin/code/output/example_figure.(out|res)'
│ might be helpful to understand and solve the issue.
│ 
│ Relevant pointers:
│ - evaluation of Julia code blocks: https://franklinjl.org/code/
│ 
│ Details:
│ ArgumentError: Package Distributions not found in current path:
│ - Run `import Pkg; Pkg.add("Distributions")` to install the Distributions package.
```
Can you add that to the docs project?

**Comment by greimel** [comment]
Done! I hadn't realized that the docs are actually built for non-member contributors these days. That's great news!

**Comment by greimel** [comment]
I would like to use this for a lecture in about a week.

If it's not too much of a hassle for you, I'd appreciate this being merged and tagged before May 8.

If you don't plan a new tag soon, please let me know and I'll adjust my material accordingly

**Comment by SimonDanisch** [comment]
I'll try to release a new version!

**Comment by SimonDanisch** [comment]
Would a breaking version work for you as well?

**Comment by greimel** [comment]
that's fine, thanks for asking!

**Pull request merged**